### ISSUE TITLE
Requiring a model

### ISSUE THREAD

**Issue by bwestergard** [balderdashy/sails#2985]
Let's say I have a module, `x.js`, whose functionality I'd like to define in terms of a sails model (created with `sails generate api cats`). How do I include the fully hydrated `Cats` model object (e.g. `Cats.findOne()`) in `x.js`?

**Comment by bwestergard** [comment]
I suppose one workaround is, when calling `x.js` from (e.g.) a controller, to pass the model in to a functiont that it exports. This obviously isn't ideal. I guess what I'm looking for is a function I can call like:

```javascript
var Cats = require('sails').HydrateModel(require('../api/models/Cats.js'));
```

[https://github.com/balderdashy/sails/blob/7c7ff990b25af3fa3d51ec97fe9ad26d9362bc91/lib/hooks/orm/build-orm.js#L61][This function] looks close, but doesn't get exported.

**Comment by nikhiljohn10** [comment]
With that bash command, you create a Cats.js in models. In sails, Models are global. Create a function inside /api/models/Cats.js 
```
module.exports = {

  attributes: {
     // Model attributes here
  },

  feedCat : function(catID, cb ){
      Cats.findOne({ id : catID }).exec(cb);
  }
};
```

here is your x.js

```
var x = require('x');
var catID = x.getID('kitty');
Cats.feedCat(catID, function(err, cat){
  if(err) console.log("Bad cat.");
  else console("Feeding cat ",cat.name);
});
```

**Comment by bwestergard** [comment]
Thanks for the prompt reply. Let me rephrase.

Let's say the module is `/lib/x.js`, that is, a plain old nodejs module that isn't loaded through the sails process (so the fact that Cats is "global" in that sails-specific sense doesn't help here). How do I include `Cats`?

I'd like to do this because all of the static code analysis tools (linting, dependency charting, etc) and testing libraries (e.g. proxyquire) are built around the assumption that code is factored into commonjs modules that require other commonjs modules.

**Comment by nikhiljohn10** [comment]
try `npm link /lib/x.js` then it will install the module locally in sails. Then u can use require as normal.

Ref: http://stackoverflow.com/questions/8088795/installing-a-local-module-using-npm

**Comment by bwestergard** [comment]
I want to require the model in `x.js`.

**Comment by nikhiljohn10** [comment]
Yes. When the module is installed in node_module folder, you can use `require()` in models function too. Try it. This is normal node case.

**Comment by tjwebb** [comment]
@bwestergard 

> Let's say the module is /lib/x.js, that is, a plain old nodejs module that isn't loaded through the sails process (so the fact that Cats is "global" in that sails-specific sense doesn't help here). How do I include Cats?

It's not loaded through the sails process, but you're trying to have it behave as a sails model? That's a pretty strange requirement, and suggests that something might be wrong with your application design. The only way I know of to accomplish this would require a bunch of complex acrobatics and hackery.

You also might be trying to do this backwards. Instead of writing tons of code in your Model, just write that code in a separate node module, and require that module from your Model and from wherever else you need it. Then trying to contrive a hydrated sails model outside of sails is not necessary.

**Comment by bwestergard** [comment]
@tjwebb To phrase it another way: am I to understand that sails users are encouraged to structure their code by relying on "the sails process" exclusively?

**Comment by bwestergard** [comment]
After reading through some of the code, I've decided what I really want to do is convert the sailsjs app I've inherited into one that just uses waterline.

**Comment by nikhiljohn10** [comment]
I was able to call `Cats.feed()` inside x.js. But x.js must be inside the project folder.
`var x = require('./../../lib/x.js');`
So u can use the above code to access the lib.

#### All you need to do is keep the x.js somewhere inside project folder.

**Comment by kevinburkeshyp** [comment]
Not actually a crazy requirement... we run into this a ton when trying to run tests. Sometimes you don't want/need to load the entire world. I think what @bwestergard is getting at is, what is the function that changes a model definition (object with list of keys, values) to add on the helpers (`findOne`, `find`, `update`, etc).

**Comment by tjwebb** [comment]
> we run into this a ton when trying to run tests. Sometimes you don't want/need to load the entire world. 

Object mocking is usually the solution for this type of thing. You could also just disable all the sails hooks besides the `orm` hook, and run `sails.load`. Then you'll only get the Waterline bindings, but in still in a bona fide sails environment.

> am I to understand that sails users are encouraged to structure their code by relying on "the sails process" exclusively?

I don't know what this means. Usually when people say this, they have a particular way of doing things, and get upset when sails has a different way of doing things, and the two methods clash in some unfortunate way. Ultimately, it's just a decision: you can bend toward the framework, or you can bend the framework toward you. I find that the first approach is almost always easier. 

That said, Sails allows you to hack it however much you want: at the end of the day, it's still javascript, where you can stomp on whatever functions you want. You can override and disable hooks, and well as implement your own. It's extremely flexible. But the farther you stray from convention, the harder it becomes to untangle problems and find support from the community.
